### PR TITLE
Max length for naming

### PR DESCRIPTION
**Pull request by 0del**
TODO list:
- [x] Apply max length validation for temporary name
- [x] Operations validation
- [x] Schema name validation
- [x] triggers name validation

fixes: https://github.com/xataio/pgroll/issues/64

### CODE REVIEW

File: pkg/migrations/op_set_fk.go [hunk 1]
```diff
@@ -149,6 +149,13 @@ func (o *OpSetForeignKey) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpSetForeignKey) Validate(ctx context.Context, s *schema.Schema) error {
+	name := TriggerName(o.Table, TemporaryName(o.Column))
+	if len(name) > MaxNameLength {
+		return InvalidNameLengthError{
+			Identity: name,
+			Max:      MaxNameLength,
+		}
+	}
```
Comment (andrew-farries) [inline pkg/migrations/op_set_fk.go:158]: I don't think it makes sense to check the names of the triggers. The trigger functions are internal to `pgroll` and their names won't mean anything to the users starting the migration.

I would remove all the checks for trigger name length and just check the names of the user-defined parts of migrations like column and constraint names.

**Comment by SferaDev** [comment]
@andrew-farries Updated with main and applied the changes to make CI checks pass

File: pkg/migrations/errors.go [hunk 2]
```diff
@@ -183,6 +183,15 @@ func (e InvalidReplicaIdentityError) Error() string {
 	return fmt.Sprintf("replica identity on table %q must be one of 'NOTHING', 'DEFAULT', 'INDEX' or 'FULL', found %q", e.Table, e.Identity)
 }
 
+type InvalidNameLengthError struct {
+	Name string
+	Max  int
+}
+
+func (e InvalidNameLengthError) Error() string {
+	return fmt.Sprintf("max length of %q is %d", e.Name, e.Max)
```
Comment (andrew-farries) [inline pkg/migrations/errors.go:192]:
```suggestion
	return fmt.Sprintf("length of %q (%d) exceeds maximum length of %d", e.Name, len(e.Name), e.Max)
```

File: pkg/migrations/name.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+func validateName(name string) error {
+	if len(name) > MaxNameLength {
+		return InvalidNameLengthError{Name: name, Max: MaxNameLength}
+	}
+	return nil
+}
+
+func ValidateName(name string) error {
+	return validateName(name)
+}
```
Comment (andrew-farries) [inline pkg/migrations/name.go:17]: Get rid of `validateName` and use the exported `ValidateName` everywhere.

File: pkg/migrations/op_change_type.go [hunk 4]
```diff
@@ -138,5 +138,10 @@ func (o *OpChangeType) Validate(ctx context.Context, s *schema.Schema) error {
 	if o.Down == "" {
 		return FieldRequiredError{Name: "down"}
 	}
+
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
+
```
Comment (andrew-farries) [inline pkg/migrations/op_change_type.go:145]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_drop_column.go [hunk 5]
```diff
@@ -56,6 +56,9 @@ func (o *OpDropColumn) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpDropColumn) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
```
Comment (andrew-farries) [inline pkg/migrations/op_drop_column.go:61]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_drop_constraint.go [hunk 6]
```diff
@@ -122,6 +122,10 @@ func (o *OpDropConstraint) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpDropConstraint) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
+
```
Comment (andrew-farries) [inline pkg/migrations/op_drop_constraint.go:128]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_drop_not_null.go [hunk 7]
```diff
@@ -130,6 +130,10 @@ func (o *OpDropNotNull) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpDropNotNull) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
+
```
Comment (andrew-farries) [inline pkg/migrations/op_drop_not_null.go:136]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_set_check.go [hunk 8]
```diff
@@ -144,6 +144,10 @@ func (o *OpSetCheckConstraint) Rollback(ctx context.Context, conn *sql.DB) error
 }
 
 func (o *OpSetCheckConstraint) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
+
```
Comment (andrew-farries) [inline pkg/migrations/op_set_check.go:150]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_set_fk.go [hunk 9]
```diff
@@ -145,6 +145,9 @@ func (o *OpSetForeignKey) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpSetForeignKey) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
```
Comment (andrew-farries) [inline pkg/migrations/op_set_fk.go:150]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_set_notnull.go [hunk 10]
```diff
@@ -162,6 +162,10 @@ func (o *OpSetNotNull) Rollback(ctx context.Context, conn *sql.DB) error {
 }
 
 func (o *OpSetNotNull) Validate(ctx context.Context, s *schema.Schema) error {
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
+
```
Comment (andrew-farries) [inline pkg/migrations/op_set_notnull.go:168]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/migrations/op_set_unique.go [hunk 11]
```diff
@@ -150,6 +150,9 @@ func (o *OpSetUnique) Validate(ctx context.Context, s *schema.Schema) error {
 	if o.Name == "" {
 		return FieldRequiredError{Name: "name"}
 	}
+	if err := validateName(o.Column); err != nil {
+		return err
+	}
```
Comment (andrew-farries) [inline pkg/migrations/op_set_unique.go:155]: I don't think it makes sense to check the length of the column name on the operation here as it must already exist in the database. The check can be removed.

File: pkg/roll/execute.go [hunk 12]
```diff
@@ -112,6 +112,9 @@ func (m *Roll) StartDDLOperations(ctx context.Context, migration *migrations.Mig
 func (m *Roll) ensureViews(ctx context.Context, schema *schema.Schema, version string) error {
 	// create schema for the new version
 	versionSchema := VersionedSchemaName(m.schema, version)
+	if err := migrations.ValidateName(versionSchema); err != nil {
+		return err
+	}
```
Comment (andrew-farries) [inline pkg/roll/execute.go:117]: This is too late to check the length of the new version schema - at this point a potentially lengthy migration start operation has already taken place so it's a poor user experience for the migration to fail with a validation check at this stage. I'd remove the check here and if we need it we can add a check earlier in the migration start process as a separate PR.

**Review by andrew-farries — CHANGES_REQUESTED** [review]

**Comment by andrew-farries** [comment]
Bump @0del 👋 

This PR is almost ready to go; it just needs some removals.

**Pull request closed without merging**